### ISSUE TITLE
[debops.slapd] Configure TLS certificates - apparmor getting in the way

### ISSUE THREAD

**Issue by mikenowak** [debops/debops#456]
While running this role with pretty much default settings I came across this error

```
TASK [debops.slapd : Configure TLS certificates (first time only)] 
*************************************************************************
fatal: [ldap1.internal.example.jp]: FAILED! => {"changed": true, "cmd": "ldapmodify -Y EXTERNAL -H
 ldapi:/// -f /tmp/tmp.mocM3TIqD0/configure_tlscerts.ldif", "delta": "0:00:00.012040", 
"end": "2018-09-27 18:13:30.458942", "msg": "non-zero return code", "rc": 80, 
"start": "2018-09-27 18:13:30.446902", "stderr": "SASL/EXTERNAL authentication started\nSASL
 username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth\nSASL SSF: 
0\nldap_modify: Other (e.g., implementation specific) error (80)", 
"stderr_lines": ["SASL/EXTERNAL authentication started", "SASL username: 
gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth", "SASL SSF: 0", "ldap_modify: Other (e.g., implementation specific) error (80)"],
 "stdout": "modifying entry \"cn=config\"", "stdout_lines": ["modifying entry \"cn=config\""]}
```

It appears that apparmor is getting in the way - perhaps an area for improvement?

```
Sep 27 18:13:30 ldap1 kernel: audit: type=1400 audit(1538061210.453:22): apparmor="DENIED"
 operation="open" profile="/usr/sbin/slapd" name="/etc/pki/realms/domain/private/key.pem" 
pid=17221 comm="slapd" requested_mask="r" denied_mask="r" fsuid=111 ouid=0
```

For the record this is on Ubuntu 18.04 and I don't believe this is officinally supported yet.

**Comment by drybjed** [comment]
I'd like to add AppArmor support in DebOps soon. Ubuntu already has it, and Debian Buster will most likely have it enabled by default. One thing that I have to do first is to fix [this `debops.grub` issue](https://github.com/debops/ansible-grub/issues/8#issuecomment-312424389), then the AppArmor role will be able to use `debops.grub` to enable AppArmor support.

I'm not sure yet how the rest of the roles will have to work - can you just drop AppArmor rules in `/etc/apparmor.d/` and have them applied automatically, or do you need to run a command to have them evaluated? Basically I'm not sure if the AppArmor role would need to be a dependency of all the roles that provide their own rules, or not. I guess we will see.

**Comment by mikenowak** [comment]
hi @drybjed ,thanks for responding!

It appears that there is a contrib role  `debops-contrib.apparmor`. I have initially missed this in my research.

I came across the usaged example in the documentation for `debops.dnsmasq`, so my understanding is that something among the lines of the below is what we need:

```
slapd__apparmor__local_dependent_config:

  'usr.sbin.slapd':
    - comment: 'Allow slapd to read ssl certificates'
      rules:
        - '/etc/pki/* r'
```

I am going to try this and report back.

**Comment by drybjed** [comment]
Yup, something like this. The current `debops-contrib.apparmor` role is a bit old though, I want to rewrite it to use the `parse_kv_items` filter which will let you put configuration in lists instead of dictionaries, eaiser to merge.

The AppArmor role will probably have to be a dependency for other roles to setup AppArmor policies beforehand. This is an incentive to keep it short (small number of tasks) and self-contained. And it should be able to be enabled/disabled globally, so that hosts without AppArmor still work.

**Comment by imrejonk** [comment]
@mikenowak did running `debops-contrib.apparmor` with the slapd configuration help? If so, we can just add the slapd__apparmor__local_dependent_config or similar and modify the `service/slapd` playbook. But maybe not before a refresh of the apparmor role.
